### PR TITLE
CDAP-6486 fix memory leak when starting programs

### PR DESCRIPTION
**Pull request by albertshau**
Adding a hack to workaround TWILL-177. Avoids creating a
ZKDiscoveryService if possible to avoid the leak.

### CODE REVIEW

**Comment by albertshau** [comment]
http://builds.cask.co/browse/CDAP-DUT4408

File: cdap-app-fabric/src/main/java/org/apache/twill/internal/AbstractTwillController.java [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+      logCancellable = kafkaClient.getConsumer().prepare()
+        .addFromBeginning(Constants.LOG_TOPIC, 0)
+        .consume(new LogMessageCallback(logHandlers));
+    }
+  }
+
+  @Override
+  public final ServiceDiscovered discoverService(String serviceName) {
+    // HACK to work around TWILL-177. CDAP never calls this method.
+    if (discoveryServiceClient == null) {
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/org/apache/twill/internal/AbstractTwillController.java:118]: this is the hack. assignment used to happen in the constructor.

**Comment by gokulavasan** [comment]
LGTM 👍

**Comment by albertshau** [comment]
after letting the master run for a while with this fix, still seeing lots of watchers and ZKClients, so holding off on merging this.

**Comment by albertshau** [comment]
ok this fixes the leak when starting programs. There may be other issues, but will handle them separately.

**Pull request merged**